app/src/main/java/org/simple/clinic/home/overdue/phonemask/PhoneMaskBottomSheetController.kt

```diff
@@ -0,0 +1,78 @@
+package org.simple.clinic.home.overdue.phonemask
+
+import io.reactivex.Observable
+import io.reactivex.ObservableSource
+import io.reactivex.ObservableTransformer
+import io.reactivex.rxkotlin.ofType
+import io.reactivex.rxkotlin.withLatestFrom
+import org.simple.clinic.ReplayUntilScreenIsDestroyed
+import org.simple.clinic.ReportAnalyticsEvents
+import org.simple.clinic.home.overdue.phonemask.CallClicked.NormalCallClicked
+import org.simple.clinic.home.overdue.phonemask.CallClicked.SecureCallClicked
+import org.simple.clinic.phone.Caller.UsingDialer
+import org.simple.clinic.phone.Caller.WithoutDialer
+import org.simple.clinic.phone.MaskedPhoneCaller
+import org.simple.clinic.util.RuntimePermissionResult
+import org.simple.clinic.util.RuntimePermissionResult.DENIED
+import org.simple.clinic.util.RuntimePermissionResult.GRANTED
+import org.simple.clinic.util.RuntimePermissionResult.NEVER_ASK_AGAIN
+import org.simple.clinic.widgets.UiEvent
+import javax.inject.Inject
+
+private typealias Ui = PhoneMaskBottomSheet
+private typealias UiChange = (Ui) -> Unit
+
+class PhoneMaskBottomSheetController @Inject constructor(
+    private val phoneCaller: MaskedPhoneCaller
+) : ObservableTransformer<UiEvent, UiChange> {
+
+  override fun apply(events: Observable<UiEvent>): ObservableSource<UiChange> {
+    val replayedEvents = ReplayUntilScreenIsDestroyed(events)
+        .compose(ReportAnalyticsEvents())
+        .replay()
+
+    return Observable.mergeArray(
+        requestCallPermission(replayedEvents),
+        makeCall(replayedEvents)
+    )
+  }
+
+  private fun requestCallPermission(events: Observable<UiEvent>) =
+      events
+          .ofType<CallClicked>()
+          .map { Ui::requestCallPermission }
+
+  private fun makeCall(events: Observable<UiEvent>) =
+      callPermissionResult(events)
+          .withLatestFrom(callClicked(events), patientPhoneNumberStream(events))
+          .flatMapCompletable { (permissionResult, callClicked, phoneNumber) ->
+            chooseTypeOfCall(callClicked)(phoneNumber, caller(permissionResult))
```

Comment:
I like the functional approach you're taking here. Exploiting the fact that both `normalCall` and `maskedCall` have the same signature and that Kotlin's type inference will detect this.